### PR TITLE
Improve type conversions

### PR DESCRIPTION
**Pull request by LeonMatthesKDAB**
- Fix return-by-reference conversions.
- Better error messages.
- Explanation in the book.

### CODE REVIEW

File: book/src/concepts/type-conversions.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+Before starting to define your own type converter, make sure this is actually what you want!!!
+Is the C++ compiler complaining about two types that you think conversion should be possible between?
+If not, you might have a bug in your `cxx_qt::bridge`.
+Double-check that you know where the error originates from and that the types are correct.
+
+However, if you're confident that there is a conversion missing, consider [opening an issues on the CXX-Qt repo](https://github.com/KDAB/cxx-qt/issues/new).
+Then you can use template specialization on the `cxx_qt_convert` struct to provide your own type conversion that CXX-Qt will use.
+
+Example:
+``` c++
```
Comment (ahayzen-kdab) [inline book/src/concepts/type-conversions.md:34]: should this have `c++,ignore` ?

File: book/src/concepts/type-conversions.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    }
+}
+
+}
+}
+```
+
+Then make sure to `include!` it in your cxx_qt::bridge.
+E.g.:
+``` rust
```
Comment (ahayzen-kdab) [inline book/src/concepts/type-conversions.md:51]: should this have `rust,ignore` ?

Comment (LeonMatthesKDAB) [reply book/src/concepts/type-conversions.md:51]: Probably, yes, will fix

File: book/src/concepts/type-conversions.md (hunk 1 above)
Comment (LeonMatthesKDAB) [reply book/src/concepts/type-conversions.md:34]: I don't think that makes sense for c++, as the doc tests don't compile those anyway.

**Review by ahayzen-kdab — APPROVED** [review]
LGTM, waiting for CI...

File: book/src/concepts/type-conversions.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+In most cases, you'll therefore use a UniquePtr in Rust when you need ownership of such an object.
+
+However, Qt often requires values to be returned by-value, especially when compatibility with QML is required.
+CXX-Qt therefore allows you to automatically convert certain Rust types to certain C++ types.
+
+For example, CXX-Qt can convert a `std::unique_ptr<QColor>` to a `QColor` for you.
+
+CXX-Qt by default provides conversions for:
+
+| From                  | To            |                                           |
```
Comment (Be-ing) [inline book/src/concepts/type-conversions.md:21]: `From C++ type` | `To Rust type`?

**Comment by Be-ing** [comment]
thanks for this documentation

File: book/src/concepts/type-conversions.md (hunk 3 above)
Comment (ahayzen-kdab) [reply book/src/concepts/type-conversions.md:21]: So for these cases `std::unique_ptr<T>` is the Rust type, and `T` is the C++/Qt type. And it's used for unique_ptr<T> -> T conversions.

But the types don't have to be for "Rust" and "C++", they can be for "C++" and "C++" and could be used elsewhere in our C++ generation or someone elses code.  So they can be generic from T to R as well.

**Pull request merged**